runtime/compiler/env/DependencyTable.cpp

```diff
@@ -0,0 +1,169 @@
+/*******************************************************************************
+ * Copyright IBM Corp. and others 2024
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+
+#include "control/CompilationThread.hpp"
+#include "env/DependencyTable.hpp"
+#include "env/PersistentCHTable.hpp"
+#include "env/J9SharedCache.hpp"
+
+TR_AOTDependencyTable::TR_AOTDependencyTable(TR_J9SharedCache *sharedCache) :
+   _sharedCache(sharedCache),
+   _tableMonitor(TR::Monitor::create("JIT-AOTDependencyTableMonitor")),
+   _offsetMap(decltype(_offsetMap)::allocator_type(TR::Compiler->persistentAllocator()))
+   { }
+
+void
+TR_AOTDependencyTable::classLoadEvent(TR_OpaqueClassBlock *clazz, bool isClassLoad, bool isClassInitialization)
+   {
+   auto ramClass = (J9Class *)clazz;
+
+   uintptr_t classOffset = TR_SharedCache::INVALID_ROM_CLASS_OFFSET;
+   if (!_sharedCache->isClassInSharedCache(clazz, &classOffset))
+      return;
+
+   // We only need to check if clazz matches its cached version on load; on
+   // initialization, it will be in the _offsetMap if it did match.
+   if (isClassLoad && !_sharedCache->classMatchesCachedVersion(ramClass, NULL))
+      return;
+
+   OMR::CriticalSection cs(_tableMonitor);
+
+   auto entry = getOffsetEntry(classOffset, isClassLoad);
+   if (isClassLoad)
+      entry->_loadedClasses.insert(ramClass);
+   }
+
+OffsetEntry *
+TR_AOTDependencyTable::getOffsetEntry(uintptr_t offset, bool create)
+   {
+   auto it = _offsetMap.find(offset);
+   if (it != _offsetMap.end())
+      return &it->second;
+
+   if (create)
+      {
+      PersistentUnorderedSet<J9Class *> loadedClasses(PersistentUnorderedSet<J9Class *>::allocator_type(TR::Compiler->persistentAllocator()));
+      return &(*_offsetMap.insert(it, {offset, {loadedClasses}})).second;
+      }
+
+   return NULL;
+   }
+
+void
+TR_AOTDependencyTable::invalidateUnloadedClass(TR_OpaqueClassBlock *clazz)
+   {
+   if (!_sharedCache)
```

Comment:
Sorry, I forgot to remove that. I originally initialized the dependency table with `_sharedCache(NULL)` and had that indicate that dependency tracking wasn't active, but I've since changed that.